### ISSUE TITLE
The equation of computing the gradient, what is the mechanism ? the gradient computation method is not clear in the paper, which is this function in bnn_layers.py line 74 "def update_grad(self, wreal, dwbin):" , can you show me the mechanism, thx a lot!

### ISSUE THREAD

**Issue by hezrein** [Jiaolong/xnor-net#2]

**Comment by Jiaolong** [comment]
You can find the equation from the paper (in the middle of page 6):
<a href="https://www.codecogs.com/eqnedit.php?latex=\frac{\partial{C}}{\partial{W_i}}&space;=&space;\frac{\partial{C}}{\partial{\widetilde{W_i}}}&space;(\frac{1}{n}&space;&plus;&space;\frac{\partial{sign}}{\partial{W_i}}\alpha)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\frac{\partial{C}}{\partial{W_i}}&space;=&space;\frac{\partial{C}}{\partial{\widetilde{W_i}}}&space;(\frac{1}{n}&space;&plus;&space;\frac{\partial{sign}}{\partial{W_i}}\alpha)" title="\frac{\partial{C}}{\partial{W_i}} = \frac{\partial{C}}{\partial{\widetilde{W_i}}} (\frac{1}{n} + \frac{\partial{sign}}{\partial{W_i}}\alpha)" /></a>

This is because:

<a href="https://www.codecogs.com/eqnedit.php?latex=\widetilde{W_i}&space;=&space;sign(W_i)&space;\\&space;\alpha&space;=&space;\frac{1}{n}\|W\|_{l_1}&space;\\&space;W_i&space;=&space;\alpha&space;\widetilde{W_i}&space;=&space;\frac{sign(W_i)}{n}\|W\|_{l_1}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\widetilde{W_i}&space;=&space;sign(W_i)&space;\\&space;\alpha&space;=&space;\frac{1}{n}\|W\|_{l_1}&space;\\&space;W_i&space;=&space;\alpha&space;\widetilde{W_i}&space;=&space;\frac{sign(W_i)}{n}\|W\|_{l_1}" title="\widetilde{W_i} = sign(W_i) \\ \alpha = \frac{1}{n}\|W\|_{l_1} \\ W_i = \alpha \widetilde{W_i} = \frac{sign(W_i)}{n}\|W\|_{l_1}" /></a>

**Comment by hezrein** [comment]
Thank you for your answer, but i am confused with the first term in the parenthesis, which is  1/n , how to obtain this term?

**Comment by Jiaolong** [comment]
<a href="https://www.codecogs.com/eqnedit.php?latex=W&space;\in&space;\mathbb{R}^{c\times&space;w&space;\times&space;h}&space;\\&space;n&space;=&space;c&space;\times&space;w&space;\times&space;h" target="_blank"><img src="https://latex.codecogs.com/gif.latex?W&space;\in&space;\mathbb{R}^{c\times&space;w&space;\times&space;h}&space;\\&space;n&space;=&space;c&space;\times&space;w&space;\times&space;h" title="W \in \mathbb{R}^{c\times w \times h} \\ n = c \times w \times h" /></a>

You can find it right above equation (2) in the [paper](https://arxiv.org/abs/1603.05279).

**Comment by hezrein** [comment]
thank you, but i get this: 
\frac{\partial{C}}{\partial{W_i}} =\frac{1}{n}\frac{\partial{C}}{\partial{\\||W\|_{l_1}}} + \frac{\partial{C}}{\partial{\widetilde{W_i}}}\frac{\partial{sign}}{\partial{W_i}}\alpha
how to tackle the first term, can you help me

**Comment by hezrein** [comment]
thank you, but i get this: 



how to tackle the first term, can you help me

**Comment by hezrein** [comment]
Thank you, but i got this:
<script type="text/javascript" async src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-MML-AM_CHTML">

![snap18](https://user-images.githubusercontent.com/37463977/40228861-147d08c0-5ac5-11e8-8492-ae800177790e.jpg)

can you show me how to tackle the first term, appreciate!

**Comment by hezrein** [comment]
<script type="text/javascript" async src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-MML-AM_CHTML">

**Comment by hezrein** [comment]
<a href="http://www.codecogs.com/eqnedit.php?latex=\frac{\partial{C}}{\partial{W_i}}&space;=&space;\frac{1}{n}\frac{\partial{C}}{\partial{\\||W\|_{l_1}}}&space;&plus;&space;\frac{\partial{C}}{\partial{\widetilde{W_i}}}\frac{\partial{sign}}{\partial{W_i}}\alpha" target="_blank"><img src="http://latex.codecogs.com/gif.latex?\frac{\partial{C}}{\partial{W_i}}&space;=&space;\frac{1}{n}\frac{\partial{C}}{\partial{\\||W\|_{l_1}}}&space;&plus;&space;\frac{\partial{C}}{\partial{\widetilde{W_i}}}\frac{\partial{sign}}{\partial{W_i}}\alpha" title="\frac{\partial{C}}{\partial{W_i}} = \frac{1}{n}\frac{\partial{C}}{\partial{\\||W\|_{l_1}}} + \frac{\partial{C}}{\partial{\widetilde{W_i}}}\frac{\partial{sign}}{\partial{W_i}}\alpha" /></a>

**Comment by hezrein** [comment]
i got this, it is not same as the paper:
<a href="http://www.codecogs.com/eqnedit.php?latex=\frac{\partial{C}}{\partial{W_i}}&space;=&space;\frac{1}{n}\frac{\partial{C}}{\partial{\\||W\|_{l_1}}}&space;&plus;&space;\frac{\partial{C}}{\partial{\widetilde{W_i}}}\frac{\partial{sign}}{\partial{W_i}}\alpha" target="_blank"><img src="http://latex.codecogs.com/gif.latex?\frac{\partial{C}}{\partial{W_i}}&space;=&space;\frac{1}{n}\frac{\partial{C}}{\partial{\\||W\|_{l_1}}}&space;&plus;&space;\frac{\partial{C}}{\partial{\widetilde{W_i}}}\frac{\partial{sign}}{\partial{W_i}}\alpha" title="\frac{\partial{C}}{\partial{W_i}} = \frac{1}{n}\frac{\partial{C}}{\partial{\\||W\|_{l_1}}} + \frac{\partial{C}}{\partial{\widetilde{W_i}}}\frac{\partial{sign}}{\partial{W_i}}\alpha" /></a>

**Comment by hezrein** [comment]
i got this, it is not same as the paper:

<a href="http://www.codecogs.com/eqnedit.php?latex=\frac{\partial{C}}{\partial{W_i}}&space;=&space;\frac{1}{n}\frac{\partial{C}}{\partial{\\||W\|_{l_1}}}&space;&plus;&space;\frac{\partial{C}}{\partial{\widetilde{W_i}}}\frac{\partial{sign}}{\partial{W_i}}\alpha" target="_blank"><img src="http://latex.codecogs.com/gif.latex?\frac{\partial{C}}{\partial{W_i}}&space;=&space;\frac{1}{n}\frac{\partial{C}}{\partial{\\||W\|_{l_1}}}&space;&plus;&space;\frac{\partial{C}}{\partial{\widetilde{W_i}}}\frac{\partial{sign}}{\partial{W_i}}\alpha" title="\frac{\partial{C}}{\partial{W_i}} = \frac{1}{n}\frac{\partial{C}}{\partial{\\||W\|_{l_1}}} + \frac{\partial{C}}{\partial{\widetilde{W_i}}}\frac{\partial{sign}}{\partial{W_i}}\alpha" /></a>

**Comment by hezrein** [comment]
thank you for your reply, but i got the following expression , that is not consistent with the paper

<a href="http://www.codecogs.com/eqnedit.php?latex=\frac{\partial{C}}{\partial{W_i}}&space;=&space;\frac{{\widetilde{W_i}}}{n}\frac{\partial{C}}{\partial{\\||W\|_{l_1}}}&space;&plus;&space;\frac{\partial{C}}{\partial{\widetilde{W_i}}}\frac{\partial{sign}}{\partial{W_i}}\alpha" target="_blank"><img src="http://latex.codecogs.com/gif.latex?\frac{\partial{C}}{\partial{W_i}}&space;=&space;\frac{{\widetilde{W_i}}}{n}\frac{\partial{C}}{\partial{\\||W\|_{l_1}}}&space;&plus;&space;\frac{\partial{C}}{\partial{\widetilde{W_i}}}\frac{\partial{sign}}{\partial{W_i}}\alpha" title="\frac{\partial{C}}{\partial{W_i}} = \frac{{\widetilde{W_i}}}{n}\frac{\partial{C}}{\partial{\\||W\|_{l_1}}} + \frac{\partial{C}}{\partial{\widetilde{W_i}}}\frac{\partial{sign}}{\partial{W_i}}\alpha" /></a>
